ic-storage/src/stable.rs

```diff
@@ -0,0 +1,390 @@
+//!
+//!     fn upgrade(previous: Self::Previous) -> Self {
+//!         Second(format!("{}, {}", previous.0, previous.1))
+//!     }
+//! }
+//! ```
+//!
+//! ## Reading
+//!
+//! Read the latest implementation from stable memory.
+//! If there is a previous version stored, this version will be upgraded.
+//!
+//! ```
+//! use ic_storage::stable::{Versioned, read};
+//! # use ic_cdk::export::candid::CandidType;
+//! # use serde::Deserialize;
+//!
+//! # #[derive(Debug, Deserialize, CandidType)]
+//! # struct First(usize, usize);
+//! # impl Versioned for First {
+//! #     type Previous = ();
+//! #     fn version() -> u32 { 1 }
+//! #     fn upgrade((): ()) -> Self {
+//! #         First(0, 0)
+//! #     }
+//! # }
+//! # #[derive(Debug, Deserialize, CandidType)]
+//! # struct Second(String);
+//! # impl Versioned for Second {
+//! #     type Previous = First;
+//! #     fn version() -> u32 { 2 }
+//! #     fn upgrade(previous: Self::Previous) -> Self {
+//! #         Self(format!("{}, {}", previous.0, previous.1))
+//! #     }
+//! # }
+//! // #[post_upgrade]
+//! fn post_upgrade_canister() {
+//!     let second = read::<Second>().unwrap();
+//! }
+//!
+//! ```
+//!
+//! ## Writing
+//!
+//! Write the current version to stable storage.
+//!
+//! ```
+//! use ic_storage::stable::{Versioned, write};
+//! # use ic_cdk::export::candid::CandidType;
+//! # use serde::Deserialize;
+//!
+//! # #[derive(Debug, Deserialize, CandidType)]
+//! # struct First(usize, usize);
+//! # impl Versioned for First {
+//! #     type Previous = ();
+//! #     fn version() -> u32 { 1 }
+//! #     fn upgrade((): ()) -> Self {
+//! #         First(0, 0)
+//! #     }
+//! # }
+//! # fn get_first() -> First { First(1, 2) }
+//! // #[pre_upgrade]
+//! fn pre_upgrade_canister() {
+//!     let first: First = get_first();
+//!     write(&first).unwrap();
+//! }
+//! ```
+use std::mem::size_of;
+
+#[cfg(test)]
+use crate::testing::{stable_bytes, stable_read, stable_size, StableWriter};
+
+#[cfg(not(test))]
+use ic_cdk::api::stable::{stable_bytes, stable_read, stable_size, StableWriter};
+
+use ic_cdk::export::candid::de::IDLDeserialize;
+use ic_cdk::export::candid::ser::IDLBuilder;
+use ic_cdk::export::candid::types::CandidType;
+use serde::Deserialize;
+
+use crate::{Error, Result};
+
+const VERSION_SIZE: usize = size_of::<u32>();
+
+/// Versioned data that can be written to, and read from stable storage.
+pub trait Versioned: for<'de> Deserialize<'de> + CandidType {
+    /// The previous version of this data.
+    /// If there is no previous version specify a unit (`()`).
+    /// This is required until defaults for associated types are stable.
+    ///
+    /// A unit has a version number of zero, and a `Previous` type of a unit,
+    /// which means it's not possible to upgrade a unit from anything but it self,
+    /// and calling `upgrade` will simply return `()`.
+    type Previous: Versioned;
+
+    /// The version of the data
+    fn version() -> u32;
```

Comment:
Can't we provide a default implementation for this method?
```
fn version() -> u32 { Self::Previous.version() + 1 }
```
In this case manual setting of versions would only be required if we delete some older state? Setting the versions by hand seems error prone...

Reply:
That's a clever idea, nice one!
Especially since you can still delete the old state and the compiler would catch you if you didn't update any dependencies on it.

It does raise the question: what should we do in the event that an old state was removed. I will take this conversation to Slack.

Nice one, Maxim!